test/unit/org/apache/cassandra/cql3/validation/operations/AlterTest.java

```diff
@@ -269,6 +276,63 @@ public void testCreateAlterKeyspaces() throws Throwable
                                   "max_threshold", "32")));
     }
 
+    @Test
+    public void testCreateAlterKeyspacesRFWarnings() throws Throwable
+    {
+        requireNetwork();
+        Thread.sleep(1000); // Hack: requireNetwork() will clear warnings async so give some time to complete
```

Comment:
I haven't been able to reproduce this. Is it the problem that the warnings are `null` if `requireNetwork` hasn't finished, or that there are other previous contents in them?

Reply:
If you run:
- Test class + no `requireNetwrok()`: pass
- Test class + `requireNetwrok()`: fail unless you sleep
- Test method + no `requireNetwrok()`: fail bc test sees 0 nodes and we short circuit that
- Test method + `requireNetwrok()`: pass

The problem being warnings goes to null as they get 'cleared'.

Reply:
so in this case maybe we should
- poll until warnings are cleared, or
- assert that warnings are cleared after this sleep.

Reply:
The test already asserts there is only 1 warn and that's the right one correct?

Reply:
well, the difference is for a person debugging the failures:
"warnings are not cleared up in the beginning"
vs
"1 was expected and got 2"

Reply:
I disagree here. A person debugging the test will be under the impression the poll guarantees the 0 warnings which is not true. A sleep signals a _best effort_ to workaround an async issue as in other tests. The 'expected 0 warnings at the start of the test' is implicit in the assert expecting only 1 warning.

Reply:
up to you... best solution would be polling wait for zero.

Reply:
But to be honest: `sleep` is one of the most popular ways of introducing flickering in the tests.

Reply:
It is indeed. It is an anti-pattern but still a workaround until the async behavior is fixed. Polling is plain wrong imo which is worst I would say. We have 2 bad options and we need to choose one :man_shrugging:

Reply:
Polling seems like a good way to avoid the test becoming flaky. However, I'm still unable to reproduce the failure when the current wait is absent. I have removed the wait [here](https://github.com/adelapena/cassandra/commits/CASSANDRA-16296-review-3) and run both the individual test and the entire suite/class in the multiplexer:
* [method](https://jenkins-dse.build.dsinternal.org/view/Parameterized/job/parameterized-testall/690/)
* [class](https://jenkins-dse.build.dsinternal.org/view/Parameterized/job/parameterized-testall/691/)

It's 100 runs for the class and 200 runs for the method, without any failure. Running them locally a few dozen times doesn't produce any failure either.

Reply:
PS I also didn't manage to reproduce the failure locally

Reply:
Hi,

Thx Andres for multiplexing it but running the method individually won't repro the issue. I get it, sometimes, only if running the whole class. It failed only 1 out of 20 for me today whereas it used to fail every time some days ago...

As I am failing to explain why the polling is wrong let me try to show it with code assuming `while(warns not empty) sleep(1)`;

- Test method starts and 'requireNetwork()' is called
- Polling loop sees 0 warns so doesn't sleep
- Test runs CQL and a warning is generated
- Some 'requireNetwork' async call clears warnings
- Assert fails as it sees 0 warnings instead of 1

Now we do the same without polling

- Test method starts and 'requireNetwork()' is called
- Sleep(1) + some 'requireNetwork' async call clears warnings hopefully within the sleep. This is the source of flakiness
- Test runs CQL and a warning is generated
- Assert succeeds

The problem is that waiting for warns to be empty is _not_ what we need. We need to wait for warns to be stable (aka requireNetwork to have finished completely). Now you can tell me I should do instead a `do sleep(1) while(!warns.empty())` and then it will work. But what's the point in polling on a random variable? It's like polling for pears when you're waiting on apples. It's the sleep that makes things work, the poll does nothing but confuse the reader imo.

This is the best I can explain it. If I am not managing to explain it let's just remove or poll (any of both won't sleep) and then we can fix it when/if it starts failing in CI.

Reply:
> Thx Andres for multiplexing it but running the method individually won't repro the issue. I get it, sometimes, only if running the whole class. It failed only 1 out of 20 for me today whereas it used to fail every time some days ago...

The jobs above also include [100 runs](https://jenkins-dse.build.dsinternal.org/view/Parameterized/job/parameterized-testall/691/) for the whole class. [Here](https://jenkins-dse.build.dsinternal.org/view/Parameterized/job/parameterized-testall/692/) and [here](https://jenkins-dse.build.dsinternal.org/view/Parameterized/job/parameterized-testall/693/) are two more runs of the whole class with 200 repetitions each. That totals 500 runs of the whole class without a single failure, so I'd say that definitely we can't reproduce it with the multiplexer.

Unfortunately testing this with CircleCI is not that easy, but I have seen a few full unit tests runs that pass without the sleep:
* https://app.circleci.com/pipelines/github/adelapena/cassandra/175/workflows/3432945a-05bc-40b4-93cf-607b1ce6566c/jobs/1291
* https://app.circleci.com/pipelines/github/adelapena/cassandra/175/workflows/4a879181-921a-47bf-8cd3-a886fbb99fa7/jobs/1297
* https://app.circleci.com/pipelines/github/adelapena/cassandra/175/workflows/a39779a1-e059-4d5c-9563-3de78d9bd864/jobs/1303
* https://app.circleci.com/pipelines/github/adelapena/cassandra/175/workflows/544a5037-0261-49d0-b5c1-9dfdfac09bc5/jobs/1311
* https://app.circleci.com/pipelines/github/adelapena/cassandra/175/workflows/1c58b779-7d80-4e79-b7d7-1b73aecc62b8/jobs/1314

I guess that the (in)famous poll, if needed, doesn't have to be for `!warns.empty` but for whatever signals that `requireNetwork` has finished and therefore it's safe to use. Said that, if we still think that the test can fail without the wait and we don't want to spend more time investigating what's the condition that signals the finalization of `requireNetwork` I think I can live with the current sleep.

Reply:
> doesn't have to be for !warns.emptybut for whatever signals that requireNetwork has finished

That is _exactly_ the point I was trying to make. Thx for summarizing it. I am happy to either remove the Sleep or leave it. You guys let me know what you prefer.